File: 9/python/rope.py
#!/usr/bin/env python3

import sys
from numpy import array

rope_len = int(sys.argv[1])
rope = [array([0, 0]) for i in range(rope_len)]
visited = {tuple(rope[-1])}
move_map = {
    "D": array([0, -1]),
    "U": array([0, 1]),
    "L": array([-1, 0]),
    "R": array([1, 0]),
}

for line in sys.stdin:
    orient, steps = line.split(" ")
    for i in range(int(steps)):
        rope[0] += move_map[orient]
        for knot in range(1, len(rope)):
            dt = rope[knot - 1] - rope[knot]
            if 2 < sum(dt ** 2):
                rope[knot] += (dt > 0).astype(int) - (dt < 0)
            else:
                break
        visited.add(tuple(rope[-1]))
print(len(visited))

File: 9/rust/rope.rs
use std::{io, io::prelude::*};
extern crate ndarray;
use ndarray::{arr1};


fn main() {
    let rope_len = std::env::args().nth(1).unwrap().parse().unwrap();
    let mut rope = Vec::new();
    for _ in 0..rope_len {
        rope.push(arr1(&[0, 0]));
    }
    let mut visited = std::collections::HashSet::new();
    visited.insert(rope[rope.len() - 1].clone());

    for line in io::stdin().lock().lines() {
        let line = line.unwrap();
        let mut iter = line.split(" ");
        let orient = iter.next().unwrap();
        let steps: i32 = iter.next().unwrap().parse().unwrap();
        for _ in 0..steps {
            match orient {
                "D" => rope[0][1] -= 1,
                "U" => rope[0][1] += 1,
                "L" => rope[0][0] -= 1,
                "R" => rope[0][0] += 1,
                &_ => todo!(),
            }
            for knot in 1..rope.len() {
                let mut dt = &rope[knot - 1] - &rope[knot];
                if 2 < (&dt * &dt).sum() {
                    dt.mapv_inplace(|e: i32| if e.abs() > 1 {e / 2} else {e});
                    rope[knot] = &rope[knot] + &dt;
                } else {
                    break;
                }
            }
            visited.insert(rope[rope.len() - 1].clone());
        }
    }
    println!("{}", visited.len());
}
